tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,376 @@
+use indexmap::IndexMap;
+use log::{debug, trace};
+use std::hash::Hash;
+use tokio_sync::oneshot;
+use tower_service::Service;
+use tower_util::Ready;
+
+/// Caches services, segregating ready and unready services.
+#[derive(Debug)]
+pub struct ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+{
+    next_ready_index: Option<usize>,
+    ready: IndexMap<K, S>,
+    pending: stream::FuturesUnordered<Pending<K, S, Req>>,
+    cancelations: IndexMap<K, oneshot::Sender<()>>,
+}
+
+#[derive(Debug)]
+enum PendingError<K, E> {
+    Canceled(K),
+    Inner(K, E),
+}
+
+/// A Future that becomes satisfied when an `S`-typed service is ready.
+///
+/// May fail due to cancelation, i.e. if the service is removed from discovery.
+#[derive(Debug)]
+struct Pending<K, S, Req> {
+    key: Option<K>,
+    cancel: oneshot::Receiver<()>,
+    ready: tower_util::Ready<S, Req>,
+}
+
+// === ReadyCache ===
+
+impl<K, S, Req> Default for ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+    S: Service<Req>,
+{
+    fn default() -> Self {
+        Self {
+            next_ready_index: None,
+            ready: IndexMap::default(),
+            cancelations: IndexMap::default(),
+            pending: stream::FuturesUnordered::new(),
+        }
+    }
+}
+
+impl<K, S, Req> ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+{
+    /// Returns the total number of services in the cache.
+    pub fn len(&self) -> usize {
+        self.ready_len() + self.pending_len()
+    }
+
+    /// Returns the number of services in the ready set.
+    pub fn ready_len(&self) -> usize {
+        self.ready.len()
+    }
+
+    /// Returns the number of services in the unready set.
+    pub fn pending_len(&self) -> usize {
+        self.pending.len()
+    }
+
+    /// Returns true iff the given key is in the unready set.
+    pub fn pending_contains<Q: Hash + Equivalent<K>>(&self, key: &Q) -> bool {
+        self.cancelations.contains_key(key)
+    }
+
+    /// Obtains a reference to a service in the ready set by key.
+    pub fn get_ready<Q: Hash + Equivalent<K>>(&self, key: &Q) -> Option<(usize, &K, &S)> {
+        self.ready.get_full(key)
+    }
+
+    /// Obtains a mutable reference to a service in the ready set by key.
+    pub fn get_ready_mut<Q: Hash + Equivalent<K>>(
+        &mut self,
+        key: &Q,
+    ) -> Option<(usize, &K, &mut S)> {
+        self.ready.get_full_mut(key)
+    }
+
+    /// Obtains a reference to a service in the ready set by index.
+    pub fn get_ready_index(&self, idx: usize) -> Option<(&K, &S)> {
+        self.ready.get_index(idx)
+    }
+
+    /// Obtains a mutable reference to a service in the ready set by index.
+    pub fn get_ready_index_mut(&mut self, idx: usize) -> Option<(&mut K, &mut S)> {
+        self.ready.get_index_mut(idx)
+    }
+
+    /// Evicts an item from the cache.
+    pub fn evict<Q: Hash + Equivalent<K>>(&mut self, key: &Q) -> bool {
+        let canceled = if let Some(c) = self.cancelations.swap_remove(key) {
```

Comment:
I'd love a small comment here about what eviction does. Specifically, something that points out that there are two cases: the service is in the pending set (not ready), or in the ready set. If the former, we cancel it, and it will be evicted when it becomes ready (why?). If the latter, we just evict it immediately.